### PR TITLE
Make _relval GTs symbolic and resnapshot Run3 data GTs

### PR DESCRIPTION
**Pull request by tvami**
#### PR description:

Discussed in https://cms-talk.web.cern.ch/t/reviewing-auto-relval-content-and-making-it-symbolic-gt/19584
We make the `_relval` GTs symbolic since they change less often and this way we can avoid mismatches between the GT and the GTs meant to be used with a fixed L1T menu.

We also propose to move away from using the offline (rereco) GT for data validations in the relvals, they should use the prompt GT instead (which in Run-3 is certainly almost already physics quality).

To be consistent with the relval GT, the frozen GTs were resnapshoted as well. The diff in GTs is just in the visible IOVs:
 * https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/126X_dataRun3_HLT_frozen_v1/126X_dataRun3_HLT_frozen_v2
 * https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/126X_dataRun3_Express_frozen_v1/126X_dataRun3_Express_frozen_v2
 * https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/126X_dataRun3_Prompt_frozen_v1/126X_dataRun3_Prompt_frozen_v2
 * https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/126X_dataRun3_v1/126X_dataRun3_v2

There is a lot of things using `auto:run3_data`
https://github.com/cms-sw/cmssw/search?p=2&q=auto%3Arun3_data
which we should probably revise, but that is beyond this PR.

#### PR validation:

```
test parameters:
  - workflows = 139.004
 ```
runs fine.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not a backport. Given the GT names and that the current data taking release is 12_6_X, it would make sense to backport to 12_6_X

### CODE REVIEW

**Comment by tvami** [comment]
test parameters:
  - workflows = 139.004

**Comment by tvami** [comment]
@sunilUIET are the relvals at pre-releases taking a wf number from CMSSW, or does that run through some other system?

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by sunilUIET** [comment]
> 

Hi @tvami , as far as I understand so far, WF numbers are coming from CMSSW as defined in relval_standard/upgrade.py. If there are some hidden steps, I let @kskovpen comment.

**Comment by missirol** [comment]
@tvami

 - I guess `L1Menu_Collisions2015_25nsStage1_v5L` is just a typo (it's `L1Menu_Collisions2015_25nsStage1_v5`).

 - In order to make the addOnTests work, please update `autoCondHLT.py` as follows:
   ```bash
   git checkout CMSSW_13_0_0_pre3 "${CMSSW_BASE}"/src/Configuration/HLT/python/autoCondHLT.py
   sed -i "s|_relval'|'       |g" "${CMSSW_BASE}"/src/Configuration/HLT/python/autoCondHLT.py
   sed -i "s|run1_hlt'       |run1_hlt_relval'|g" "${CMSSW_BASE}"/src/Configuration/HLT/python/autoCondHLT.py
   sed -i "s|run2_hlt'       |run2_hlt_relval'|g" "${CMSSW_BASE}"/src/Configuration/HLT/python/autoCondHLT.py
   ```

Additional questions:

 - Why isn't `run2_hlt_relval` made 'symbolic' as well?

 - Is `run1_hlt_relval` only used in [`autoCondHLT.py#L44`](https://github.com/cms-sw/cmssw/blob/3e10dc1e28a9dc04b6f07c1832f20519714de354/Configuration/HLT/python/autoCondHLT.py#L44). If so, and if it is okay with AlCa, you could remove [`run1_hlt_relval` from `autoCond.py`](https://github.com/cms-sw/cmssw/blob/6caf45172ea3d835de401c49bc0aa46a0e30505f/Configuration/AlCa/python/autoCond.py#L106), and change `run1_hlt_relval` to `run2_hlt_relval` in `autoCondHLT.py` (or to `run2_hlt` if that is added to make `run2_hlt_relval` 'symbolic').

**Comment by missirol** [comment]
> is this really intended? There is no run3_hlt_relval like this

Yes, that is the idea. What `autoCondHLT.py` effectively does is to customise the GT "X" with a fixed L1T menu decided by TSG, so "X" can just be `run3_hlt` as long as the difference between `run3_hlt` and `run3_hlt_relval` is only the L1T-menu tag.

The other questions hint at the fact that, if some updates are done upstream (e.g. adding `run2_hlt`), we could remove all mentions of `_relval` in `autoCondHLT.py`. (not saying it's a big gain, it just seems clearer to me that way)

**Comment by tvami** [comment]
both `hlt_mc_GRun` and `139.004` pass in my local tests, I'll trigger tests here now

**Comment by tvami** [comment]
urgent
 * I believe it would be beneficial to have the `pre4` already use the prompt GTs instead of the rereco ones, otherwise the final release version will have widespread differences which wont come from the release changes, let me know if you agree @sunilUIET

File: Configuration/AlCa/python/autoCond.py [hunk 1]
```diff
@@ -27,24 +27,18 @@
...
-    'run3_hlt_relval'              : '126X_dataRun3_HLT_relval_v2',
-    # GlobalTag for Run3 data relvals (express GT) - identical to 126X_dataRun3_Express_v1 but with snapshot at 2023-01-06 18:15:00 (UTC)
-    'run3_data_express'            : '126X_dataRun3_Express_frozen_v1',
-    # GlobalTag for Run3 data relvals (prompt GT) - identical to 126X_dataRun3_Prompt_v1 but with snapshot at 2023-01-06 18:15:00 (UTC)
-    'run3_data_prompt'             : '126X_dataRun3_Prompt_frozen_v1',
-    # GlobalTag for Run3 offline data reprocessing - snapshot at 2023-01-06 18:15:00 (UTC)
-    'run3_data'                    : '126X_dataRun3_v1',
-    # GlobalTag for Run3 data relvals: allows customization to run with fixed L1 menu
-    'run3_data_relval'             : '126X_dataRun3_relval_v2',
+    'run2_hlt'                     : '123X_dataRun2_HLT_relval_v3',
```
Comment (missirol) [inline Configuration/AlCa/python/autoCond.py:33]: Not a problem for HLT per se, but isn't confusing that now `run2_hlt` points to a "relval" GT ? (I was expecting a frozen, non-relval, HLT GT instead)

If it's easier/clearer , there is no problem keeping `run2_hlt_relval`.

File: Configuration/HLT/python/autoCondHLT.py [hunk 2]
```diff
@@ -15,8 +15,8 @@
 #and leave unchanged the snapshot times for the other tuples.
 
 l1Menus= {
-    'Fake'         : ( ','.join( [ 'L1GtTriggerMenu_L1Menu_Collisions2012_v3_mc'             , l1MenuRecord,connectionString, l1MenuLabel, "2015-10-26 12:00:00.000"] ), ),
-    'Fake1'        : ( ','.join( [ 'L1Menu_Collisions2015_25nsStage1_v5'                     , l1MenuRecord,connectionString, l1MenuLabel, "2015-10-26 12:00:00.000"] ), ),
+    'Fake'         : ( ','.join( [ 'L1GtTriggerMenu_L1Menu_Collisions2012_v3_mc'             ,l1MenuRecord,connectionString, l1MenuLabel, "2015-10-26 12:00:00.000"] ), ),
+    'Fake1'        : ( ','.join( [ 'L1Menu_Collisions2015_25nsStage1_v5'                     ,l1MenuRecord,connectionString, l1MenuLabel, "2015-10-26 12:00:00.000"] ), ),
```
Comment (missirol) [inline Configuration/HLT/python/autoCondHLT.py:19]: Please undo this change, the extra whitespace is intentional.

File: Configuration/HLT/python/autoCondHLT.py [hunk 3]
```diff
@@ -41,27 +41,27 @@
...
+    'run1_data_Fake'         : ('run2_data'               ,l1Menus['Fake']),
+    'run2_data_Fake'         : ('run2_data'               ,l1Menus['Fake']),
+    'run2_data_Fake1'        : ('run2_data'               ,l1Menus['Fake1']),
+    'run2_data_Fake2'        : ('run2_data'               ,l1Menus['Fake2']),
+    'run3_data_FULL'         : ('run3_data_prompt'        ,l1Menus['FULL']),
+    'run3_data_GRun'         : ('run3_data_prompt'        ,l1Menus['GRun']),
+    'run3_data_2022v15'      : ('run3_data_prompt'        ,l1Menus['2022v15']),
+    'run3_data_HIon'         : ('run3_data_prompt'        ,l1Menus['HIon']),
+    'run3_data_PIon'         : ('run3_data_prompt'        ,l1Menus['PIon']),
+    'run3_data_PRef'         : ('run3_data_prompt'        ,l1Menus['PRef']),
```
Comment (missirol) [inline Configuration/HLT/python/autoCondHLT.py:64]: Why is this change (`run3_data` -> `run3_data_prompt`) useful here ?

GTs like `run3_data_GRun` are mainly used in HLT tests that run HLT+RECO (so, HLT and offline reconstruction in the same job; afaik, no production wf runs HLT+RECO in one go, but it is a functionality that HLT tries to maintain). I don't know if there is a realistic scenario where we'd run HLT+RECO on real data: if there is one, what would the most appropriate GT be (`run3_data` or `run3_data_prompt`)?

(If the answer is not obvious, I suggest to leave `run3_data` and HLT can take care of changing this later with a backport if needed.)

File: Configuration/AlCa/python/autoCond.py (hunk 1 above)
Comment (tvami) [reply Configuration/AlCa/python/autoCond.py:33]: it's just that `run2_hlt` is not consumed anywhere as a frozen, non-relval GT, we dont even have a queue for it, so it would have been some extra effort to do something that's then not used anywhere. You guys are the real users of this, so I can change it as you wish. Should I revert to `run2_hlt_relval` then @missirol ?

File: Configuration/HLT/python/autoCondHLT.py (hunk 2 above)
Comment (tvami) [reply Configuration/HLT/python/autoCondHLT.py:19]: ok, the lines below didn't have a whitespace so I thought this shouldn't either, I'll change it back

File: Configuration/HLT/python/autoCondHLT.py (hunk 3 above)
Comment (tvami) [reply Configuration/HLT/python/autoCondHLT.py:64]: `run3_data` points to the ReRECO GT, so should be used only once we cut a rereco GT. When we do that it will have the snapshot time, and that will not change for another year or so, on the other hand the prompt GT we resnapshot more often (like every month or so). I think using the `run3_data_prompt` is the appropriate one for HLT tests too, the `run3_data` would use the last visible conditions only (i.e. if you run this on data from the upcoming MWGR, it will use conditions from December 2022, since that's when we made the rereco GT)

**Comment by tvami** [comment]
@cmsbuild , please abort

File: Configuration/AlCa/python/autoCond.py (hunk 1 above)
Comment (missirol) [reply Configuration/AlCa/python/autoCond.py:33]: >  Should I revert to run2_hlt_relval then ?

Let's do that, to avoid unnecessary changes. Sorry for the back and forth.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by missirol** [comment]
+hlt

**Comment by kskovpen** [comment]
+pdmv

**Comment by AdrianoDee** [comment]
+1

File: Configuration/AlCa/python/autoCondModifiers.py [hunk 4]
```diff
@@ -96,3 +96,32 @@ def autoCond2017ppRef5TeV(autoCond):
...
     return autoCond
+
+def autoCondRelValForRun2(autoCond):
+
+    GlobalTagRelValForRun2 = {}
+    L1GtTriggerMenuForRelValForRun2 =    ','.join( ['L1Menu_Collisions2015_25nsStage1_v5' , "L1GtTriggerMenuRcd",             connectionString, "", "2023-01-28 12:00:00.000"] )
+    L1TUtmTriggerMenuForRelValForRun2 =  ','.join( ['L1Menu_Collisions2018_v2_1_0-d1_xml' , "L1TUtmTriggerMenuRcd",           connectionString, "", "2023-01-28 12:00:00.000"] )
+
+    for key,val in autoCond.items():
+        if 'run2_data' in key or 'run2_hlt' in key:
```
Comment (francescobrivio) [inline Configuration/AlCa/python/autoCondModifiers.py:107]: I think `if 'run2_data' in key` will generate unwanted symbolic GTs since in autoCond we also have:
 - run2_data_HEfail --> will generate `run2_data_HEfail_relval` 
 - run2_data_promptlike_hi --> will generate `run2_data_promptlike_hi_relval`

File: Configuration/AlCa/python/autoCondModifiers.py [hunk 5]
```diff
@@ -96,3 +96,32 @@ def autoCond2017ppRef5TeV(autoCond):
...
+    return autoCond
+
+def autoCondRelValForRun3(autoCond):
+
+    GlobalTagRelValForRun3 = {}
+    L1GtTriggerMenuForRelValForRun3 =    ','.join( ['L1Menu_Collisions2015_25nsStage1_v5' , "L1GtTriggerMenuRcd",             connectionString, "", "2023-01-28 12:00:00.000"] )
+    L1TUtmTriggerMenuForRelValForRun3 =  ','.join( ['L1Menu_Collisions2022_v1_4_0-d1_xml' , "L1TUtmTriggerMenuRcd",           connectionString, "", "2023-01-28 12:00:00.000"] )
+
+    for key,val in autoCond.items():
+        if 'run3_data' in key or 'run3_hlt' in key :
```
Comment (francescobrivio) [inline Configuration/AlCa/python/autoCondModifiers.py:121]: same comment here given that in autoCond we have:
 - run3_data_express --> will generate `run3_data_express_relval`
 - run3_data_prompt --> will generate `run3_data_prompt_relval`

**Comment by francescobrivio** [comment]
+alca

 - after private discussion with @cms-sw/alca-l2 and given the urgency (we would like to have this in pre4) we decided that the cleanup of the unwanted symbolic GTs can be done later (see https://github.com/cms-AlCaDB/AlCaTools/issues/85 opened as reminder). Also, given the fact that these symbolic GTs should be explicitly called somewhere (and they aren't) they remain totally harmless.

**Comment by perrotta** [comment]
"auto:run2_data_relval" is still needed by
- Configuration/PyReleaseValidation/python/relval_steps.py
- CalibTracker/SiPixelQuality/test/step3_SiPixelCalZeroBias_RAW2DIGI_RECO_ALCA.py

**Comment by tvami** [comment]
@perrotta that's fine, we are still providing `auto:run2_data_relval`, just in a different way. The wf `139.004` is also usig a _relval GT, and as you can see it runs just fine.

**Comment by perrotta** [comment]
+1
